### PR TITLE
Migrate as much as possible to pyproject.toml

### PR DESCRIPTION
**Pull request by vyasr**
This moves as much packaging information and linter configuration out of setup.[py|cfg] files into pyproject.toml with the aim of migrating to more modern Python builders in the future.

### CODE REVIEW

**Comment by vyasr** [comment]
@rlratzel @jakirkham the conda split recipe workaround being used for the service build is causing problems here. I'm fairly certain that the `--no-index` argument to pip is causing it to not use the index even for installing build dependencies (the stuff under build-system in pyproject.toml). I've tried adding them to the conda recipe in case that just works but I'm expecting it to still fail without further changes. Should we be building those packages with build isolation turned off? I believe that (in conjunction with the recipe changes) would fix the problem since then the package would be built in the environment created by conda-build that would contain the requirements specified in the recipe.

**Comment by vyasr** [comment]
Yep build isolation did it. Worked before because there was no pyproject.toml file at all so it wasn't using a PEP 518 build at all.

**Comment by jakirkham** [comment]
Yeah disabling build isolation makes sense

Normally conda-build [passes this option to `pip`]( https://github.com/conda/conda-build/blob/6aafd4a370b093ca22cf6fe7d52b2927ef049d47/conda_build/build.py#L2584-L2590 ), but the recipe here runs into an edge case where these flags need to be passed manually ( https://github.com/conda/conda-build/issues/3993 )

**Review by raydouglass — APPROVED** [review]

File: python/cugraph-dgl/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
```
[the remark this answers is not in the document]
Comment (jakirkham) [reply python/cugraph-dgl/pyproject.toml:1]: Or just 2023 since it was added this year?

[the remark this answers is not in the document]
Comment (vyasr) [reply python/cugraph-dgl/pyproject.toml:1]: Yup will switch to 2023.

I'm honestly not sure what the answer is for where we need the full header and where we don't. My general observation in RAPIDS is that we include the full header in C++/CMake files and only the one-liner in Python and associated files. I don't know if that is by design or just by happenstance.

File: python/pylibcugraph/pyproject.toml [hunk 2]
```diff
@@ -13,3 +13,46 @@ requires = [
...
 ]
 build-backend = "setuptools.build_meta"
+
+[tool.pytest.ini_options]
+testpaths = ["pylibcugraph/tests"]
+
+[project]
+name = "pylibcugraph"
+version = "23.04.00"
+description = "cuGraph - RAPIDS GPU Graph Analytics"
```
[the remark this answers is not in the document]
Comment (vyasr) [reply python/pylibcugraph/pyproject.toml:23]: I like it. I had copied what was in setup.py, but this is definitely better.

File: python/cugraph-dgl/pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (rlratzel) [reply python/cugraph-dgl/pyproject.toml]: ok, thanks. I filed [this issue](https://github.com/rapidsai/cugraph/issues/3340) so we don't forget to follow-up.

**Pull request merged**